Add winner tool

jstarry opened the pull request:
#### Problem
Need transparent deterministic tooling for determining the winners of TdS

#### Summary of Changes
* Create a new CLI tool which behaves similarly to ledger-tool for the purpose of computing the TdS prize winners. This new tool will be moved to https://github.com/solana-labs/tour-de-sol once required runtime changes are shipped to crates.io
* Compute the winners for the "Most Rewards Earned" category by summing the balances of the vote and stake accounts which are attributed to a particular validator. Fixes: https://github.com/solana-labs/solana/issues/6077
* Compute the winners for the "Confirmation Latency" category by processing the ledger entry by entry and tracking the order of votes received. Slow voters get docked a point, fast voters earn a point. Fixes: https://github.com/solana-labs/solana/issues/6078
* Compute the winners for the "Highest Availability" category by tracking voting and leader history. Missed leader slots are weighted higher than missed votes. Fixes: https://github.com/solana-labs/solana/issues/6080

#### Reference
https://forums.solana.com/t/tour-de-sol-stage-1-preliminary-compensation-design/79

Fixes https://github.com/solana-labs/solana/issues/6076

jstarry commented:
@mvines I've written the tool to output the winners (top 3 + percentile buckets), but it might just be simpler to have it write a couple CSVs with an ordered list of results for each category. Thoughts?

mvines commented on .travis.yml:
```diff
@@ -0,0 +1,7 @@
+  - rustup component add rustfmt
+  - cargo build --verbose --all
+  - cargo test --all
+  - cargo fmt --all -- --check
```
no clippy?

mvines commented on prize-tool/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+edition = "2018"
+name = "solana-tds-prize-tool"
+description = "Solana Tour de SOL prize tool"
+version = "0.20.0"
```
might as well start at 1.0.0 here, there's no tie in with the main repo version really

mvines commented on prize-tool/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+[dependencies]
+clap = "2.33.0"
+log = "0.4.8"
+solana-core = { git = "https://github.com/jstarry/solana", branch = "blocktree-process-opts" }
```
crates.io work now?

jstarry replied:
Soon, have to finish up this PR first: https://github.com/solana-labs/solana/pull/6248

jstarry replied:
Issue here: https://github.com/solana-labs/tour-de-sol/issues/16

mvines commented on prize-tool/src/main.rs:
```diff
@@ -0,0 +1,124 @@
+                .help("Starting balance of validators at the beginning of TdS"),
+        )
+        .arg(
+            Arg::with_name("solana_baseline")
```
solana_baseline -> baseline_validator perhaps?  Baseline might not be a "solana" validator

mvines commented on prize-tool/src/main.rs:
```diff
@@ -0,0 +1,124 @@
+        exit(1);
+    });
+
+    let blocktree = match Blocktree::open(&ledger_path) {
```
unwrap_or_else() here too?

CriesofCarrots replied:
@jstarry Did you decide not to do this for some reason?

jstarry replied:
😱 so sorry! Not sure how this happened, fix here: https://github.com/solana-labs/tour-de-sol/pull/17

mvines commented on prize-tool/src/confirmation_latency.rs:
```diff
@@ -0,0 +1,216 @@
+
+// Track voter latency by sequentially processing the blockchain specified by `block_slots` and
+// checkpointing the voter record after each entry.
+fn track_voter_latency(
```
This fn looks like it could turn into a maintenance burden as it replicates a bunch of code from within `process_blocktree()`.  Could add a new option for a callback were `process_blocktree()` can call us at the right time to collect the vote info?

CriesofCarrots commented on prize-tool/src/utils.rs:
```diff
@@ -0,0 +1,96 @@
+) -> Vec<(String, Vec<Winner>)> {
+    let find_bucket_index = |value: f64| -> usize {
+        let mut index = 0;
+        while index < results.len() && results[index].1 > value {
```
I think it would be a bit more idiomatic and clear to use an iterator. So the entirety of the closure would be:
```
let find_bucket_index = |value: f64| -> usize {
        results
            .iter()
            .position(|&result| result.1 <= value)
            .unwrap_or(results.len())
    };
```

CriesofCarrots commented on prize-tool/src/prize.rs:
```diff
@@ -0,0 +1,17 @@
+use solana_sdk::pubkey::Pubkey;
```
Why is this module called "prize"?
Seems like it should be "winner" or "winners"

CriesofCarrots commented on prize-tool/src/main.rs:
```diff
@@ -0,0 +1,141 @@
+        )
+        .arg(
+            Arg::with_name("starting_balance")
+                .long("starting_balance")
```
We typically use kebab case for long arguments

CriesofCarrots commented on prize-tool/src/main.rs:
```diff
@@ -0,0 +1,141 @@
+                .value_name("PUBKEY")
+                .takes_value(true)
+                .required(true)
+                .help("Public key of the baseline validator"),
```
Did you know about this? https://github.com/solana-labs/solana/blob/c2761a1259c193db112a8f2a027c5076650c54fa/cli/src/input_validators.rs#L5

You can add `.validator(is_pubkey)` and save yourself the check and exit below. It's a little nicer UX, imo.

It might be a pain to import solana-cli, although it would allow you to use the input_parsers as well.
But even copy-pasta-ing that method would be a little improvement, I think.

jstarry replied:
Didn't know about `validator(..)` thanks!

CriesofCarrots commented on prize-tool/src/availability.rs:
```diff
@@ -0,0 +1,221 @@
+//! Calculates the winners of the "Highest Availability" category in Tour de SOL by dividing the
+//! credits of validator vote accounts by (the total number of blocks in the chain) + (the number
```
I think these parens aren't quite right.
Shouldn't it be `(the total number of blocks in the chain + (the number of leader slots the validator missed * MISSED_LEADER_SLOT_WEIGHT))`?

jstarry replied:
Yes, it should be. Sorry about that!

CriesofCarrots commented on prize-tool/src/rewards_earned.rs:
```diff
@@ -0,0 +1,290 @@
+    let mut voter_stake_sum: HashMap<Pubkey, u64> = HashMap::new();
+    for (_key, account) in stake_accounts {
+        if let Some(StakeState::Stake(_authorized, _lockup, stake)) = StakeState::from(&account) {
+            let stake_sum = voter_stake_sum
```
This would be really nice as:
```
voter_stake_sum
  .entry()
  .modify(...)
  .or_insert(...);
```

CriesofCarrots commented on prize-tool/src/rewards_earned.rs:
```diff
@@ -0,0 +1,290 @@
+            let voter_stake_reward = voter_stake_rewards.remove(&voter_key).unwrap_or_default();
+
+            let validator_id = vote_state.node_pubkey;
+            if let Some(validator_reward) = validator_reward_map.get_mut(&validator_id) {
```
See above. You can use the same `.entry()` trick here.

CriesofCarrots commented on prize-tool/src/availability.rs:
```diff
@@ -0,0 +1,221 @@
+    let mut validator_credits = HashMap::new();
+    for (_voter_key, (_stake, account)) in vote_accounts {
+        if let Some(vote_state) = VoteState::from(&account) {
+            if let Some(credits) = validator_credits.get_mut(&vote_state.node_pubkey) {
```
```
validator_credits
  .entry()
  .modify(...)
  .or_insert(...);
```
?

CriesofCarrots commented on prize-tool/src/availability.rs:
```diff
@@ -0,0 +1,221 @@
+            .slot_leader_at(slot, Some(bank))
+            .unwrap();
+        if let Some(leader_stat) = validator_leader_stats.get_mut(&leader) {
+            leader_stat.total_slots += 1;
```
Same `.entry()` trick would be nice here.

CriesofCarrots commented on prize-tool/src/utils.rs:
```diff
@@ -0,0 +1,96 @@
+
+/// Returns an ordered list of slots for the blockchain ending with `last_block` and starting with
+/// `first_block`
+pub fn block_slots(first_block: Slot, last_block: Slot, blocktree: &Blocktree) -> Vec<Slot> {
```
I find this method name a little confusing, since what you mean is block_lineage.
Lineage sounds pretentious, though. Maybe block_ancestors (even though it's potentially bounded)?

jstarry replied:
How about `block_chain`?

CriesofCarrots replied:
Hahaha, I thought of that too! I didn't suggest it because I thought the term might be a little overloaded...? On the other hand, it does seem to be essentially correct.

jstarry replied:
lol I had it originally but ditched it for the same reason. I think I'll stick with it though since it's the least confusing

CriesofCarrots commented on prize-tool/src/availability.rs:
```diff
@@ -0,0 +1,221 @@
+    let num_validators = results.len();
+    let num_winners = min(num_validators, 3);
+    let baseline =
+        baseline_credits as f64 / (baseline_leader_stat.missed_slots + total_credits) as f64;
```
Shouldn't the baseline_leader_stat.missed_slots be multiplied by MISSED_LEADER_SLOT_WEIGHT as well?

jstarry replied:
Yes, that's a bug, thank you!

CriesofCarrots commented on prize-tool/src/confirmation_latency.rs:
```diff
@@ -0,0 +1,149 @@
+        let vote_state = VoteState::from(&account).unwrap();
+        let voter_entry = voter_record.remove(&voter_key).unwrap();
+        // It's possible that there are multiple vote accounts attributed to a validator
+        //   so use the max score when duplicates are found
```
This seems inconsistent with the handling in rewards_earned, where multiple vote account rewards are summed.
What are the circumstances in which we envision validators ending up with multiple vote accounts?

jstarry replied:
Yeah the inconsistency is no good, thanks for calling it out. My understanding is that we don't envision it happening but we currently don't prevent it from happening so I opted to handle it just in case.

That being said, here are some scenarios:
1. A validator could restart their node with a different vote account
2. A validator could disable the check that the vote account `node_pubkey` matches their node's pubkey. That way you could have two validator's colluding to inflate their earnings.
3. A validator could modify their code to run duplicate votes across multiple accounts

I was originally only thinking of the first scenario but I think now that to protect against gaming the winner calculations, we shouldn't be cumulating rewards across vote accounts. What do you think about using `max` for both "Confirmation Latency" and for "Highest Rewards Earned"?

CriesofCarrots replied:
I like using `max` for both for now.
I'll think on it some more, though. Scenario 1 isn't really malicious behavior, and seems relatively likely to happen.

jstarry replied:
Yeah, seems like a reasonable mistake to make. Sounds good 👍

jstarry commented:
Addressed feedback but have been investigating why my recent tests have not been seeing the expected number of vote and stake accounts from the bank. ~Will fix tests following that investigation!~ Fixed

EDIT: the reason was because I was testing the tool after `cargo install`ing it and that doesn't work because the blocktree processor doesn't have the right paths to load the native programs 🤦‍♂

jstarry commented:
@CriesofCarrots @mvines  I've addressed feedback, added more tests, and verified the tool against a local cluster. Please take another look when you have time!

mvines commented on winner-tool/src/main.rs:
```diff
@@ -0,0 +1,179 @@
+//! This tool calculates the quantitative category winners for Tour de SOL.
+//!
+//! NOTE: Ledger processing uses native programs, so this tool must be invoked with `cargo run`.
```
Can you please move this info, and other general operating instructions into a new section of https://github.com/solana-labs/tour-de-sol/blob/master/README-internal.md

mvines replied:
Oh and one better (fine as a follow-up PR if that's easier): a new shell script like https://github.com/solana-labs/tour-de-sol/blob/master/bench-tps.sh that runs the tool with all the right settings against the actual TdS cluster.

jstarry replied:
Great call https://github.com/solana-labs/tour-de-sol/issues/13

mvines commented on winner-tool/src/main.rs:
```diff
@@ -0,0 +1,179 @@
+
+    let ledger_path = PathBuf::from(value_t_or_exit!(matches, "ledger", String));
+    let starting_balance_sol = value_t_or_exit!(matches, "starting_balance", f64);
+    let baseline_id_string = value_t_or_exit!(matches, "baseline_validator", String);
```
Does this work?  `let baseline_id_string = value_t_or_exit!(matches, "baseline_validator", Pubkey);`

jstarry replied:
Yes, fixed here: https://github.com/solana-labs/tour-de-sol/pull/15/files#diff-69c184a9c7bce2c16d29851280d961faR82

mvines commented on winner-tool/src/main.rs:
```diff
@@ -0,0 +1,179 @@
+    let excluded_set = {
+        let mut set = HashSet::new();
+        set.insert(baseline_id);
+        set.insert(bootstrap_id);
```
I think we should rename `--bootstrap-leader` to `--exclude XXX`, and support multiple excludes.  Solana runs a several nodes currently in the TdS cluster, we'd want to exclude them all

jstarry replied:
https://github.com/solana-labs/tour-de-sol/issues/14

jstarry replied:
Done
